feat: Add permissions required for GitHub integration PUL-1443 (#156)
codacy/pulse-user-docs: docs/one-click-integrations/github-integration.md

-   In this case, you must send to Pulse the information about your **deployments** and the corresponding **changes** using the [Pulse CLI](../cli/cli.md) or the [Ingestion API](https://ingestion.pulse.codacy.com/v1/api-docs).

> ## Collected data {: id="collected-data"}

Reviewer: Adding this explicit ID isn't necessary because MkDocs already generates the same ID for this section title automatically. You can check that here:

https://docs.pulse.codacy.com/one-click-integrations/github-integration/#collected-data

We usually define these explicit IDs only when the section titles are very long or are prone to change (to avoid breaking incoming links).

```suggestion
## Collected data
```

<sup><span id="deployment-teams">2</span></sup>: Adding a new team or changing the composition of a team on GitHub only affects new data from that moment onwards and doesn't have an immediate impact on the dashboards. Also, deleted teams on GitHub will continue to show in Pulse.

## Which permissions does Pulse need from GitHub? {: id="gh-permissions"}

> Pulse requests only the necessary permissions from GitHub to collect your organization changes and deployment data and [keeps your information secure](https://security.codacy.com/). See below the detailed list of permissions.

Reviewer: The current version seems to indicate that Pulse collects two types of data:

- Organization changes
- Deployment data

Something along these lines would be more focused on the main data that Pulse collects to calculate the metrics, which are changes and deployments:

```suggestion
Pulse requests only the necessary permissions from GitHub to collect changes and deployment data from the repositories in your organization and [keeps your information secure](https://security.codacy.com/). See below the detailed list of permissions.
```

I don't think we need to mention here all the details that are listed in the table (which would include things like the members and teams in the organization, for example).

<table>
  <colgroup>
    <col width="20%"/>
    <col width="20%"/>
    <col width="60%"/>
  </colgroup>
  <thead>
    <tr>
      <th>Scope</th>
      <th>Permissions</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td colspan="3"><strong>Repository permissions:</strong></td>
    </tr>
    <tr>
      <td>Pull requests</td>
      <td>Read</td>

>       <td>Pulse retrieves pull request information to calculate several of the metrics presented on the dashboards. <a href="#collected-data">See the details here.</a></td>

Reviewer: Grammar improvement:

```suggestion
      <td>Pulse retrieves pull request information to calculate several metrics presented on the dashboards. <a href="#collected-data">See the details here.</a></td>
```

    </tr>
    <tr>
      <td>Contents</td>
      <td>Read</td>

>       <td>Pulse retrieves contents of files and directories in a repository to read tags and support semantic versioning.</td>

Reviewer 2: while we ask for permission to "retrieves contents of files and directories in a repository" we don't use it, and we can make it more clear to users

we only used this permission to read git tags which enable the support of the semantic version tags option in the Integration Settings. unfortunately, the permission to read git tags is the same to get access to all files of a repository

Author: Updated. See how it looks.

>       <td>Pulse retrieves tag information to support semantic versioning. For this, Pulse GitHub App requires read permission on contents of files and directories, as it's the same permission that applies to tags.</td>

Reviewer 3: ```suggestion
      <td>Pulse retrieves tag information to support semantic versioning. For this, Pulse GitHub App requires read permission on files and directories, as it's the same permission that applies to tags.</td>
```
`contents of` sounds redundant here. Read permissions already imply access to content.

Author: I wrote that way because [Contents](https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28) is in fact the GitHub object we are referring to, and also to match the term in the first column.

Do you think that it will have the same technical meaning if I remove the "contents of" as suggested, @nicklem @prcr ?

Reviewer: Hmm, I see. Then my suggestion is to use the same wording as GitHub does:

> [...] requires read permission on **the** contents of files and directories

![image](https://user-images.githubusercontent.com/60105800/205896380-8fa63c1b-9067-4627-9876-753cd4a269a6.png)

Author: Thanks! Updated.

>       <td>Pulse retrieves tag information to support semantic versioning. For this, Pulse GitHub App requires read permission on contents of files and directories, as it's the same permission that applies to tags.</td>

Reviewer: Here it's appropriate to use an article when referring to "the Pulse GitHub App". For example, you can check out [how GitHub refers to GitHub Apps](https://docs.github.com/en/developers/apps/getting-started-with-apps/about-apps#about-github-apps).

```suggestion
      <td>Pulse retrieves tag information to support semantic versioning. For this, the Pulse GitHub App requires read permission on contents of files and directories, as it's the same permission that applies to tags.</td>
```

>       <td>Pulse retrieves tag information to support semantic versioning. For this, the Pulse GitHub App requires read permission on the contents of files and directories, as it's the same permission that applies to tags.</td>

Author: @afsmeira Would "Pulse retrieves tag information **to track deployments via semantic versioning tags**." work?

>       <td>Pulse retrieves tag information to support semantic versioning. For this, the Pulse GitHub App requires read permission on the contents of files and directories, as it's the same permission that applies to tags.</td>

Reviewer 4: Yes. 👍

>       <td>Pulse retrieves tag information to support semantic versioning. For this, the Pulse GitHub App requires read permission on the contents of files and directories, as it's the same permission that applies to tags.</td>

Author: Updated!

    </tr>
    <tr>
      <td>Issues</td>
      <td>Read</td>

>       <td>Pulse retrieves issue information to get the top-level comments of pull requests.</td>

Reviewer 2: looks good. the team can review it too!

>       <td>Pulse retrieves issue information to get the top-level comments of pull requests.</td>

Author: Thanks! Changing this PR as ready for review.

    </tr>
    <tr>
      <td colspan="3"><strong>Organization permissions:</strong></td>
    </tr>
    <tr>
      <td>Webhooks</td>
      <td>Read & Write</td>

>       <td>Pulse creates webhooks for organization and repository events (...) to calculate metrics. </td>

Reviewer 2: we subscribe to the following events: "create", "pull_request", "pull_request_review"

Reference: https://docs.github.com/en/developers/webhooks-and-events/webhook-events-and-payloads

>       <td>Pulse creates webhooks for organization and repository events (...) to calculate metrics. </td>

Author: Thanks! Updated.

>       <td>Pulse creates webhooks for organization and repository events (create, pull_request, pull_request_review) to calculate metrics. </td>

Reviewer: You're absolutely right, since we're using an HTML table the Markdown syntax is no longer interpreted. We can use `<code>` tags instead:

```html
Pulse creates webhooks for organization and repository events (<code>create</code>, <code>pull_request</code>, <code>pull_request_review</code>) to calculate metrics.
```

>       <td>Pulse creates webhooks for organization and repository events (<code>create</code>, <code>pull_request</code>, <code>pull_request_review</code>) to calculate metrics.</td>

Reviewer 4: We can further unpack this paragraph for more visibility:

- organization webhooks are used to know when
    - repositories are added to or removed from our integration
    - the integration is suspended or uninstalled
- repository events
    - `create` - to know when tags are created (basically the same reason as 'Contents' above)
    - `pull_request` - to know when a pull request is created, edited or deleted (same reason as 'Pull Requests' above)
    - `pull_request_review` - to know when a review is created, edited or deleted (this is just for the 'Changes & Reviews' dashboard)
    
In general, the webhook permission and associated events are necessary to gather data in real time. Other permissions are for historical data (obtained only after the integration is created). Members permission is used for both historical and real-time.

You don't have to add all of this info, it's more of a FYI.

Author: Thank you so much for the detailed information! I updated this paragraph. See how it looks now.

>       <td>Pulse creates webhooks for organization and repository events (<code>create</code>, <code>pull_request</code>, <code>pull_request_review</code>) to calculate metrics.</td>

Author: What about "to calculate real-time metrics"?

    </tr>
    <tr>
      <td>Members</td>
      <td>Read</td>

>       <td>Pulse retrieves information about organization members and teams to enforce permissions, user management, and billing purposes.</td>

Reviewer 3: ```suggestion
      <td>Pulse retrieves information about organization members and teams to enforce permissions, manage users, and for billing purposes.</td>
```
Reword suggestion
